### PR TITLE
#316, #336, #268 crop season refactor, time series interpolation and aggregation

### PR DESCRIPTION
**Pull request by krsnapaudel**
I tried to rewrite parts of alignment and dataset.

Removing the alignment of `eos_date` to end of year makes the code clean. Now we have to make sure everything works.

### CODE REVIEW

**Comment by krsnapaudel** [comment]
@ronvree, @aikepotze : I have removed some alignment code and passed crop calendar to dataset. Would you like to come up with an outline for data processing in torch dataset? Thanks.

NOTE: Data is as is now. No end-of-season alignment to end of year. Seems to work for other models. Need to make this work for nn models.

**Comment by krsnapaudel** [comment]
With the latest update, validation of AgML workshop results works. See `agml_workshop.py`. The results are different when aligning data to crop season. This part needs more investigation.

File: cybench/datasets/torch_dataset.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+        :return: a dict containing the data sample
+        """
+        sample = super(TorchDataset, self).__getitem__(index)
+        if self._aggregate_time_series_to is not None:
+            if self._aggregate_time_series_to == "week":
+                num_time_steps = int(np.ceil(self._max_season_window_length / 7))
+            else:
+                num_time_steps = int(np.ceil(self._max_season_window_length / 10))
+
+            # make sure time series data has the same number of time steps
```
Comment (ronvree) [inline cybench/datasets/torch_dataset.py:135]: Is this not already enforced during initialization?

File: cybench/runs/agml_workshop.py [hunk 2]
```diff
@@ -17,20 +18,34 @@
...
+    CROP_CALENDAR_DOYS,
+    CROP_CALENDAR_DATES,
 )
 
 
 class LSTMModel(BaseModel, nn.Module):
     def __init__(
-        self, num_rnn_layers=1, rnn_hidden_size=64, num_outputs=1, *args, **kwargs
+        self,
+        time_series_have_same_length=False,
```
Comment (ronvree) [inline cybench/runs/agml_workshop.py:37]: doesnt the model assume this is the case? In my understanding it doesnt support an alternative. Doesnt it make this code redundant?

File: cybench/datasets/torch_dataset.py (hunk 1 above)
Comment (krsnapaudel) [reply cybench/datasets/torch_dataset.py:135]: It's not. We could do, but probably less straightforward.

File: cybench/runs/agml_workshop.py (hunk 2 above)
Comment (krsnapaudel) [reply cybench/runs/agml_workshop.py:37]: We run workshop results validation with data that does not need aggregation.

File: cybench/datasets/torch_dataset.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+            df_ts = df_ts.merge(
+                df_all_dates, how="outer", on=[KEY_LOC, KEY_YEAR, "date"]
+            )
+            del df_all_dates
+
+            # NOTE: interpolate fills data in forward direction.
+            df_ts = df_ts.sort_index().interpolate(method="linear")
+            # fill NAs in the front with 0.0
+            df_ts.fillna(0.0, inplace=True)
+
```
Comment (aikepotze) [inline cybench/datasets/torch_dataset.py:90]: I suggest decoupling interpolation from aggregation. Right now using daily data assumes data is complete and equal ts length.

I suggest moving line 50-90 to outside of the aggregation condition.

Comment (aikepotze) [reply cybench/datasets/torch_dataset.py:90]: It's also fine if we add this in the future

**Review by aikepotze — APPROVED** [review]

**Pull request merged**